### DISCUSSION TITLE
The Mystery of the Missing KeyCodes

### DISCUSSION THREAD

**Discussion by pop** [bevyengine/bevy#9465] [Q&A]
Hey all. I'm working on an app which needs a simple text-box with user input. Not a complex text editor -- just accept keystrokes into a text file and save it/load it later.

I have looked into tools like `egui`, which I believe can do the job -- but I am curious _if_ vanilla Bevy can work. I am trying to avoid adding dependencies where possible.

The road-block I am hitting is some keys do not have a [`KeyCode`](https://docs.rs/bevy/0.11.0/bevy/input/keyboard/enum.KeyCode.html), so I cannot detect that key being pressed. For example `!`, `#` and `$` are handled by [`EventReader<KeyboardInput>`](https://docs.rs/bevy/0.11.1/bevy/input/keyboard/struct.KeyboardInput.html) as `None` KeyCodes whereas `@` and all letters/numbers are handled fine.

## Example

Below is a simplified example demonstrating the problem:
```rust
use bevy::input::keyboard::KeyboardInput;
use bevy::input::ButtonState;
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, init)
        .add_systems(Update, update)
        .run();
}

fn init(mut commands: Commands) {
    // UI Camera
    commands.spawn((
        Camera2dBundle { ..default() },
        UiCameraConfig { show_ui: true },
    ));

    // Text Box
    commands.spawn(TextBundle::from_section(
        "Start typing! (Escape to clear)\n",
        TextStyle {
            color: Color::WHITE,
            ..default()
        },
    ));
}

fn update(mut query: Query<&mut Text>, mut events: EventReader<KeyboardInput>) {
    for KeyboardInput {
        key_code,
        scan_code,
        state,
        ..
    } in events.iter()
    {
        if *state == ButtonState::Pressed {
            let mut text = query.single_mut();
            if *key_code == Some(KeyCode::Escape) {
                text.sections.drain(1..);
            } else {
                let style = TextStyle {
                    color: Color::WHITE,
                    ..default()
                };
                let added = format!(
                    "Pressed: KeyCode: {:?} / ScanCode {:?}\n",
                    key_code, scan_code
                );
                text.sections.push(TextSection::new(added, style));
            }
        }
    }
}
```

## Demo

![20230816_20h52m55s_grim](https://github.com/bevyengine/bevy/assets/1897147/e45a7e1a-4c34-4b94-85dd-827bc11dab74)

In the screenshot I typed `ASDF` then `1234` then `Shift` + `1234`.

I have the [`ScanCode`](https://docs.rs/bevy/0.11.1/bevy/input/keyboard/struct.ScanCode.html) but I can't determine any way to programmatically correlate that with a keyboard character, `Shift`-ed or otherwise.

## Prior art

I'm sure as work on UI and a [Bevy Editor](https://docs.rs/bevy/0.11.1/bevy/input/keyboard/struct.ScanCode.html) picks up this will be resolved.

It also seems like [`winit` had some recent work on keyboard input event](https://github.com/rust-windowing/winit/milestone/3?closed=1) so this issue may be moot once those changes land in Bevy.

## Request

Does anybody have a workaround in the meantime? `winit` and `bevy` may update to include all of the changes mentioned above, but has anybody seen a way of handling textbox input?

## Workaround

For anybody curious my workarounds are:
* Accept a subset of keys and communicate that to the user, probably just letters and numbers for simplicity.
* For my game editor I plan to edit files in an external editor and import the files as assets. The assets should auto-update if the user edits files while the editor is running somewhat reducing development friction.
* Once the upstream `winit` changes land in Bevy I will update this thread with a new status and updated example(s).

**Comment by ThierryBerger** [comment]
YES! Thanks for asking!

In my experience, [altgr](https://github.com/mvlabat/bevy_egui/pull/149#issuecomment-1443771002) was incorrectly handled by winit, it was [fixed](https://github.com/rust-windowing/winit/pull/2662) in winit, come help me https://github.com/bevyengine/bevy/pull/8745.
